README.md
# \Term\ANSIColor

## NAME

\Term\ANSIColor - Color screen output using ANSI escape sequences.

## SYNOPSIS

```php
use \Term\ANSIColor;

require_once '/path/to/vendor/autoload.php';

// Get a colored string
echo ANSIColor::colored("Yellow text\n", 'yellow');
echo ANSIColor::colored("Normal text on magenta\n", null, 'magenta');
echo ANSIColor::colored("Underscored text\n", null, null, 'underscore');
echo ANSIColor::colored("Red on cyan\n", 'red', 'cyan');
echo ANSIColor::colored("Blinking blue text on light gray\n", 'blue', 'light_gray', 'blink');

// Get a colored string by an alias
ANSIColor::setAlias('debug', 'red', 'light_gray');
echo ANSIColor::colored("[DEBUG] Debug message in red on light_gray\n" , 'debug');
ANSIColor::setAlias('error', 'red', 'black');
echo ANSIColor::colored("[ERROR] Error message in red on black\n" , 'error');
```

## INSTALLATION

To install this package into your project via composer, add the following snippet to your `composer.json`. Then run `composer install`.

```
"require": {
    "travail/term-ansicolor": "dev-master"
}
```

If you want to install from github, add the following:

```
"repositories": [
    {
        "type": "vcs",
        "url": "git@github.com:travail/php-Term-ANSIColor.git"
    }
]
```

## METHODS

### colored

`string colored(string $text, string $foreground_color, string $background_color, string $attribute)`

Returns a colored text, or `Exception` if passed non-existent `$foreground_color`, `$background_color` or `$attribute`.

### setAlias

`array setAlias(string $alias, string $foreground_color, string $background_color, string $attribute)`

Sets an alias for the standard color and attribute.

### getAlias

`array getAlias([string $alais = ''])`

Returns an available alias as an array if `$alias` passed, all of alias if `$alias` not passed, or `Exception` if non-existent `$alias` passed.

### getForegourndColors

`array getForegroundColors(void)`

Returns the defined foreground colors as array.

```
array(
    'black'       => '0;30',
    'dark_gray'   => '1;30',
    'blue'        => '0;34',
    'light_blue'  => '1;34',
    'green'       => '0;32',
    'light_green' => '1;32',
    'cyan'        => '0;36',
    'light_cyan'  => '1;36',
    'red'         => '0;31',
    'light_red'   => '1;31',
    'purple'      => '0;35',
    'light_purple'=> '1;35',
    'brown'       => '0;33',
    'yellow'      => '1;33',
    'light_gray'  => '0;37',
    'white'       => '1;37',
);
```


### getBackgroundColors

`array getBackgroundColors(void)`

Returns the defined background colors as array.

```
array(
    'black'      => '40',
    'red'        => '41',
    'green'      => '42',
    'yellow'     => '43',
    'blue'       => '44',
    'magenta'    => '45',
    'cyan'       => '46',
    'light_gray' => '47',
);
```

### getAttributes

`array getAttributes(void)`

Returns the defined attributes as array.

```
array(
    'normal'     => '0',
    'bold'       => '1',
    'underscore' => '4',
    'blink'      => '5',
    'reverse'    => '7',
    'concealed'  => '8',
);
```

## AUTHOR

travail

## LICENSE

This library is free software. You can redistribute it and/or modify it under the same terms as PHP itself.


CHANGELOG.md
CHANGELOG
========

# 0.1.0 2014-05-01

* Composerized
* Tidy up README.md and example scripts to compliant with composer


lib/ANSIColor.php
<?php

namespace Term;

class ANSIColor
{
    /**
     * @var string The version of this package
     */
    const VERSION = '0.1.0';

    /**
     * @var array ANSI foreground color map
     */
    static private $foreground = array(
        'black'       => '0;30',
        'dark_gray'   => '1;30',
        'blue'        => '0;34',
        'light_blue'  => '1;34',
        'green'       => '0;32',
        'light_green' => '1;32',
        'cyan'        => '0;36',
        'light_cyan'  => '1;36',
        'red'         => '0;31',
        'light_red'   => '1;31',
        'purple'      => '0;35',
        'light_purple'=> '1;35',
        'brown'       => '0;33',
        'yellow'      => '1;33',
        'light_gray'  => '0;37',
        'white'       => '1;37',
    );

    /**
     * @var array ANSI background color map
     */
    static private $background = array(
        'black'      => '40',
        'red'        => '41',
        'green'      => '42',
        'yellow'     => '43',
        'blue'       => '44',
        'magenta'    => '45',
        'cyan'       => '46',
        'light_gray' => '47',
    );

    /**
     * @var array ANSI text attribute map
     */
    static private $attribute = array(
        'normal'     => '0',
        'bold'       => '1',
        'underscore' => '4',
        'blink'      => '5',
        'reverse'    => '7',
        'concealed'  => '8',
    );

    /**
     * @var array
     */
    static private $alias = array();

    /**
     * Colorlize and decorate string.
     *
     * @param string $string
     * @param string $f
     * @param string $b
     * @param string $a
     * @return string Colored string
     * @throws \Exception
     */
    static public function colored($string, $f = '', $b = '', $a = '')
    {
        if (self::validateAlias($f)) list($f, $b, $a) = self::$alias[$f];
        if (!$f && !$b && !$a) return $string;

        if ($f && !self::validateForegroundColor($f) && !self::validateAlias($f))
            throw new \Exception("No such a foreground color '$f'");
        if ($b && !self::validateBackgroundColor($b))
            throw new \Exception("No such a background color '$b'");
        if ($a && !self::validateAttribute($a))
            throw new \Exception("No such an attribute '$a'");

        $colored = '';
        if ($f)             $colored .= "\033[" . self::$foreground[$f] . "m";
        if ($b)             $colored .= "\033[" . self::$background[$b] . "m";
        if ($a)             $colored .= "\033[" . self::$attribute[$a] . "m";
        if ($f || $b || $a) $colored .= $string . "\033[0m";

        return $colored;
    }

    /**
     * Set an alias.
     *
     * @param string $alias
     * @param string $f
     * @param string $b
     * @param string $a
     * @return void
     * @throws \Exception
     */
    static public function setAlias($alias, $f = '', $b = '', $a = '')
    {
        if (!$alias)
            throw new \Exception('alias is required');
        if (!$f && !$b && !$a)
            throw new \Exception(
                'Specify at least one from foreground color, ' .
                'background color and text attribute');

        self::$alias[$alias] = array($f, $b, $a);
    }

    /**
     * Return an alias when $alias is given
     * return all alias when $alias is not given.
     *
     * @param string $alias
     * @return array
     * @throws \Exception
     */
    static public function getAlias($alias = '')
    {
        if (!$alias) {
            return self::$alias;
        }
        elseif (self::validateAlias($alias)) {
            return self::$alias[$alias];
        }
        else {
            throw new \Exception("No such an alias $alias");
        }
    }

    /**
     * Return ANSI foreground color map.
     *
     * @return array
     */
    static public function getForegroundColors()
    {
        return self::$foreground;
    }

    /**
     * Return ANSI background color map.
     *
     * @return array
     */
    static public function getBackgroundColors()
    {
        return self::$background;
    }

    /**
     * Return ANSI text attribute map.
     *
     * @return array
     */
    static public function getAttributes()
    {
        return self::$attribute;
    }

    /**
     * Validate given ANSI foreground color.
     *
     * @param string $color
     * @return boolean
     */
    static public function validateForegroundColor($color)
    {
        return isset(self::$foreground[$color]);
    }

    /**
     * Validate given ANSI background color.
     *
     * @param string $color
     * @return boolean
     */
    static public function validateBackgroundColor($color)
    {
        return isset(self::$background[$color]);
    }

    /**
     * Validate given ANSI text attribute.
     *
     * @param string $attr
     * @return boolean
     */
    static public function validateAttribute($attr)
    {
        return isset(self::$attribute[$attr]);
    }

    /**
     * Validate given alias.
     *
     * @param string $alias
     * @return boolean
     */
    static public function validateAlias($alias)
    {
        return isset(self::$alias[$alias]);
    }
}
